std/format/read.d

```diff
@@ -663,6 +664,86 @@ if (isSomeString!(typeof(fmt)))
     assert(aa2 == ["hello":1, "world":2]);
 }
 
+/**
+Reads an input range according to a format string and returns a tuple with the
+read values.
+
+Format specifiers with format character $(B 'd'), $(B 'u') and $(B
+'c') can take a $(B '*') parameter for skipping values.
+
+The second version of `formattedRead` takes the format string as
+template argument. In this case, it is checked for consistency at
+compile-time.
+
+Params:
+    Args = a variadic list of types of the arguments
+ */
+template formattedRead(Args...)
+if (Args.length && allSatisfy!(isType, Args))
+{
+    import std.typecons : Tuple;
+
+    /**
+    Params:
+        r = an $(REF_ALTTEXT input range, isInputRange, std, range, primitives),
+            where the formatted input is read from
+        fmt = a $(MREF_ALTTEXT format string, std,format)
+        Range = the type of the input range `r`
+        Char = the character type used for `fmt`
+
+    Returns:
+        A Tuple!Args with the elements filled. If the input range `r` ends early,
+        the missing arguments will be default initialized.
+
+    Throws:
+        A $(REF_ALTTEXT FormatException, FormatException, std, format)
+        if reading did not succeed.
+    */
+    Tuple!Args formattedRead(Range, Char)(auto ref Range r, const(Char)[] fmt)
+    {
+        import core.lifetime : forward;
+        Tuple!Args args;
+        .formattedRead(forward!r, fmt, args.expand);
+        return args;
+    }
+}
+
+///
+@safe pure unittest
+{
+    import std.typecons : tuple;
+
+    auto complete = "hello!34.5:124".formattedRead!(string, double, int)("%s!%s:%s");
+    assert(complete == tuple("hello", 34.5, 124));
+
+    auto missing = "hello!34.5:".formattedRead!(string, double, int)("%s!%s:%s");
+    assert(missing == tuple("hello", 34.5, int.init));
+}
+
+/// ditto
+template formattedRead(alias fmt, Args...)
+if (!isType!fmt && isSomeString!(typeof(fmt)) && Args.length && allSatisfy!(isType, Args))
+{
+    import std.typecons : Tuple;
+    Tuple!Args formattedRead(Range)(auto ref Range r)
+    {
+        import core.lifetime : forward;
+        Tuple!Args args;
+        .formattedRead!fmt(forward!r, args.expand);
+        return args;
+    }
+}
+
+/// The format string can be checked at compile-time:
+@safe pure unittest
+{
+    import std.typecons : tuple;
+
+    auto expected = tuple("hello", 124, 34.5);
+    auto result = "hello!124:34.5".formattedRead!("%s!%s:%s", string, int, double);
```

Comment:
If the format string can be checked at compile-time, why would one need to pass in types?

Reply:
The format string is not responsible for unformatting into the types the user wants, it's only to know the places in the input string that need to be unformatted and to tell the formatted the restrictions of the characters it's reading. Using `%f` can still be valid for both `float` and `double` for example, and it asserts statically when the types are not format-compatible.
```d
float value;
"123".formattedRead!"%d"(value);
// Error: static assert:  "incompatible format character for floating point argument: %d"
```

It also asserts on orphan format specifiers or if the number of arguments is higher than the required by the format string. This is useful for generic code when we don't have total control over the argument quantity or the format string. The format string might be generated during compile time causing it to assert if some malformation occurs, or the variadic template arguments might be calculated at compile time asserting for the same reasons.

Reply:
I actually think I found a bug in `checkFormatException` for `%f` specifier.
```d
int i;
"123".formattedRead!"%f"(i);
// should static assert but instead throws at run time with:
// std.format.FormatException@/dlang/dmd/linux/bin64/../../src/phobos/std/format/internal/read.d(104): Wrong unformat specifier '%f' for int
```

Reply:
> I actually think I found a bug in `checkFormatException` for the `%f` format.
> 
> ```d
> int i;
> "123".formattedRead!"%f"(i);
> // should static assert but instead throws at run time with:
> // std.format.FormatException@/dlang/dmd/linux/bin64/../../src/phobos/std/format/internal/read.d(104): Wrong unformat specifier '%f' for int
> ```

Could you please file a bug report for this?